### PR TITLE
feat: Add `chat.getURLPreview` endpoint

### PR DESCRIPTION
**Pull request by matheusbsilva137**
## Proposed changes (including videos or screenshots)

 - Add `chat.getURLPreview` endpoint to enable users to retrieve URL previews prior to sending messages.

## Steps to test or reproduce

This endpoint requires `url` and `roomId` properties.
**Example request (cURL):**
```
curl --request GET 'http://localhost:3000/api/v1/chat.getURLPreview?roomId=650b53843e77b9b901c7cba6&url=https://www.youtube.com/watch?v=no050HN4ojo' \
--header 'X-Auth-Token: AlBrFCgo7ChqzKhTP0dViC-jS7N_HL4vyQT2E-fs9x0' \
--header 'X-User-Id: iGiuqqMDCNMXDhKcY'
```

<details>
  <summary><b>Example response</b></summary>

  ```
{
  "urlPreview": {
    "url": "https://www.youtube.com/watch?v=no050HN4ojo",
    "meta": {
      "oembedUrl": "https://www.youtube.com/watch?v=no050HN4ojo",
      "oembedTitle": "Rocket.Chat - El líder de comunicación de código abierto",
      "oembedAuthorName": "Rocket.Chat",
      "oembedAuthorUrl": "https://www.youtube.com/@RocketChatApp",
      "oembedType": "video",
      "oembedVersion": "1.0",
      "oembedProviderName": "YouTube",
      "oembedProviderUrl": "https://www.youtube.com/",
      "oembedThumbnailUrl": "https://i.ytimg.com/vi/no050HN4ojo/hqdefault.jpg",
      "oembedHtml": "<iframe style=\"max-width: 100%;width:400px;height:225px\" width=\"200\" height=\"113\" src=\"https://www.youtube.com/embed/no050HN4ojo?feature=oembed\" frameborder=\"0\" allow=\"accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share\" allowfullscreen title=\"Rocket.Chat - El líder de comunicación de código abierto\"></iframe>"
    },
    "headers": {
      "contentType": "application/json"
    },
    "parsedUrl": {
      "host": "www.youtube.com",
      "hash": null,
      "pathname": "/oembed",
      "protocol": "https:",
      "port": null,
      "query": {
        "maxheight": "200",
        "url": "https://www.youtube.com/watch?v=no050HN4ojo"
      },
      "search": "?maxheight=200&url=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3Dno050HN4ojo",
      "hostname": "www.youtube.com",
      "slashes": true,
      "auth": null,
      "path": "/oembed?maxheight=200&url=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3Dno050HN4ojo",
      "href": "https://www.youtube.com/oembed?maxheight=200&url=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3Dno050HN4ojo"
    },
    "ignoreParse": true
  },
  "success": true
}
```
</details>

<details>
  <summary><b>Example request for sending a message with previously generated previews</b></summary>

 - The `parseUrls` param must be sent so that previews are not re-generated (when this param is sent, we use the previews sent to the endpoint, no new ones are generated)
  ```
curl -H "X-Auth-Token: AlBrFCgo7ChqzKhTP0dViC-jS7N_HL4vyQT2E-fs9x0" \
     -H "X-User-Id: iGiuqqMDCNMXDhKcY" \
     -H "Content-type:application/json" \
     http://localhost:3000/api/v1/chat.sendMessage \
     -d '{"message": { "rid": "GENERAL", "msg": "This is a test! https://www.youtube.com/watch?v=no050HN4ojo", "urls": [{"url":"https://www.youtube.com/watch?v=no050HN4ojo","meta":{"oembedUrl":"https://www.youtube.com/watch?v=no050HN4ojo","oembedTitle":"Rocket.Chat - El líder de comunicación de código abierto","oembedAuthorName":"Rocket.Chat","oembedAuthorUrl":"https://www.youtube.com/@RocketChatApp","oembedType":"video","oembedVersion":"1.0","oembedProviderName":"YouTube","oembedProviderUrl":"https://www.youtube.com/","oembedThumbnailUrl":"https://i.ytimg.com/vi/no050HN4ojo/hqdefault.jpg","oembedHtml":"<iframe style=\"max-width: 100%;width:400px;height:225px\" width=\"200\" height=\"113\" src=\"https://www.youtube.com/embed/no050HN4ojo?feature=oembed\" frameborder=\"0\" allow=\"accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share\" allowfullscreen title=\"Rocket.Chat - El líder de comunicación de código abierto\"></iframe>"},"headers":{"contentType":"application/json"},"parsedUrl":{"host":"www.youtube.com","hash":null,"pathname":"/oembed","protocol":"https:","port":null,"query":{"maxheight":"200","url":"https://www.youtube.com/watch?v=no050HN4ojo"},"search":"?maxheight=200&url=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3Dno050HN4ojo","hostname":"www.youtube.com","slashes":true,"auth":null,"path":"/oembed?maxheight=200&url=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3Dno050HN4ojo","href":"https://www.youtube.com/oembed?maxheight=200&url=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3Dno050HN4ojo"},"ignoreParse":true}], "parseUrls": false }}'
```
</details>

## Further comments

> **Note**
> Returned URL previews contain the `ignoreParse` property so that they are not retrieved again when attached to a message and sent.

> **Note**
> Removed attachments check on `rocketUrlParser` since the `attachment` property is not returned from `getUrlMetaWithCache` (see [this line](https://github.com/RocketChat/Rocket.Chat/blob/14fc38e60e029ddfd6c1732c80a649fd7014fdca/apps/meteor/app/oembed/server/server.ts#L124))

TC-938

### CODE REVIEW

**Review by heitortanoue — APPROVED** [review]

File: apps/meteor/app/oembed/server/server.ts [hunk 1]
```diff
@@ -151,10 +188,6 @@ const getUrlMeta = async function (
 		return;
 	}
 
-	if (content.attachments) {
```
Comment (MarcosSpessatto) [inline apps/meteor/app/oembed/server/server.ts:154]: Is this change really necessary here? I saw your comment `Removed attachments check on...` on the PR description. I believe if this is not a requirement for the PR and the feature, we should not change that now, it would be better to create another PR only with this change. WDYT?

File: apps/meteor/app/oembed/server/server.ts [hunk 2]
```diff
@@ -273,37 +306,10 @@ const rocketUrlParser = async function (message: IMessage): Promise<IMessage> {
...
-			continue;
-		}
-
-		const data = await getUrlMetaWithCache(item.url);
-
-		if (!data) {
-			continue;
-		}
-
-		if (isOEmbedUrlContentResult(data) && data.attachments) {
```
Comment (MarcosSpessatto) [inline apps/meteor/app/oembed/server/server.ts:286]: Same as before, if we are removing this only and only because it's not being returned by `getUrlMetaWithCache `, it would be better to keep as is here, and create another PR, so it's easier to track the changes.

**Review by MarcosSpessatto — CHANGES_REQUESTED** [review]

File: apps/meteor/app/oembed/server/server.ts (hunk 2 above)
Comment (matheusbsilva137) [reply apps/meteor/app/oembed/server/server.ts:286]: [This](https://github.com/RocketChat/Rocket.Chat/blob/14fc38e60e029ddfd6c1732c80a649fd7014fdca/apps/meteor/app/oembed/server/server.ts#L124) is the correct link for the line I mentioned :nerd_face:
I was just not so sure about doing that when I created this PR since doing that (keeping this line) would require me to add another param (the `attachments` array) to the `parseUrl` function just to remove it later in another PR. Does it look fine to you?

**Review by MarcosSpessatto — APPROVED** [review]

File: apps/meteor/app/api/server/v1/chat.ts [hunk 3]
```diff
@@ -822,3 +823,30 @@ API.v1.addRoute(
...
 );
+
+API.v1.addRoute(
+	'chat.getURLPreview',
+	{ authRequired: true },
+	{
+		async get() {
+			const { roomId, url } = this.queryParams;
+
+			if (!roomId) {
```
Comment (sampaiodiego) [inline apps/meteor/app/api/server/v1/chat.ts:834]: I saw you created the `ajv` schema for this endpoint but is not using it and you're checking the params by hand like this one.

I'd say to use the `ajv` function `isChatGetURLPreviewProps` to validate the endpoints params instead of checking them manually 😉

File: apps/meteor/app/api/server/v1/chat.ts [hunk 4]
```diff
@@ -822,3 +823,30 @@ API.v1.addRoute(
...
+			if (!url) {
+				throw new Meteor.Error('error-url-param-not-provided', 'The required "url" query param is missing.');
+			}
+
+			if (!(await canAccessRoomIdAsync(roomId, this.userId))) {
+				throw new Meteor.Error('error-not-allowed', 'Not allowed');
+			}
+
+			const { urlPreview } = await OEmbed.parseUrl(url);
+			urlPreview.ignoreParse = true;
```
Comment (sampaiodiego) [inline apps/meteor/app/api/server/v1/chat.ts:847]: can you explain why is this needed? or what is the purpose of this?

File: apps/meteor/app/oembed/server/server.ts [hunk 5]
```diff
@@ -128,6 +135,36 @@ const getUrlContent = async (urlObj: URL, redirectCount = 5): Promise<OEmbedUrlC
...
+				parsedUrlObject.meta.oembedHtml = insertMaxWidthInOembedHtml(parsedUrlObject.meta.oembedHtml) || '';
+			}
+		}
+
+		if (data.headers?.contentLength) {
+			parsedUrlObject.headers = { ...parsedUrlObject.headers, contentLength: data.headers.contentLength };
+		}
+
+		if (data.headers?.contentType) {
+			parsedUrlObject.headers = { ...parsedUrlObject.headers, contentType: data.headers.contentType };
```
Comment (sampaiodiego) [inline apps/meteor/app/oembed/server/server.ts:160]: even though this code was just moved from its previous location to here, but I'd say it could be refactored to be easier to read and don't do multiple spreads of `parsedUrlObject.headers` .. something like:


```ts
if (!isURL) {
  return ...;
}

if (!data) {
  return ...;
}

return {
   headers: {
    ...parsedUrlObject.headers,
    ...(data.headers?.contentLength ? { contentLength: data.headers.contentLength } : {}),
    ...(data.headers?.contentType ? { contentType: data.headers.contentType } : {}),
  }
}
```

wdyt?

**Review by sampaiodiego — CHANGES_REQUESTED** [review]

File: apps/meteor/app/api/server/v1/chat.ts (hunk 3 above)
Comment (matheusbsilva137) [reply apps/meteor/app/api/server/v1/chat.ts]: Oopsss :grimacing: thanks for noticing that. Should be fixed now

File: apps/meteor/app/oembed/server/server.ts (hunk 5 above)
Comment (matheusbsilva137) [reply apps/meteor/app/oembed/server/server.ts]: Good idea. I changed the final return statement to be more clear as you suggested
I had to add some extra checks, but it definitely looks clearer

File: apps/meteor/app/api/server/v1/chat.ts (hunk 4 above)
Comment (matheusbsilva137) [reply apps/meteor/app/api/server/v1/chat.ts:839]: The idea for adding this param here is that we don't parse the same URL again when the output of this endpoint is "attached" to a message and sent to RC.
Since the final purpose of this endpoint is being used by our client to generate previews before sending a message, I thought it'd make sense to assign this param here so we don't parse this URL again later when the message is sent (I felt like this should be this endpoint's responsibility, but please tell me if you think otherwise).

**Review by sampaiodiego — APPROVED** [review]
